pkg/streamingpromql/engine_fuzz_test.go

```diff
@@ -0,0 +1,270 @@
+func doComparisonQuery(t *testing.T, input TestQueryInput, ctx TestQueryContext, prepareQueriesFunc InvokeQueryEnginePreparation) {
+
+	prep := prepareQueriesFunc(input)
+
+	// Note - prometheus validates time ranges and steps in http api handler, so it will not return an error at query preparation time
+	// Note - max resolution size (end-start/step) is validated in codec.go so we don't validate that here
+	if prep.prom.error != nil || (!prep.isInstantQuery() && input.startEpoch > input.endEpoch) {
+		require.NotNil(t, prep.mimir.error)
+		return
+	}
+
+	var prom, mimir *promql.Result
+
+	// Execute our queries
+	defer prep.prom.query.Close()
+	prom = prep.prom.query.Exec(context.Background())
+
+	defer prep.mimir.query.Close()
+	mimir = prep.mimir.query.Exec(context.Background())
+
+	// Replace series Metric which are nil with an empty Labels{}
+	FixUpEmptyLabels(prom)
+
+	// Special case for posting errors which are formatted differently in Prometheus and Mimir
+	fixUpAndAssertPostingErrors(t, prom, mimir)
+
+	// Assert that the results match exactly. Note that annotations comparison will be ignored if the query is in the given ignore map
+	testutils.RequireEqualResults(t, input.query, prom, mimir, ctx.ignoreAnnotations[input.query])
+}
+
+// This test function will run like a normal unit test with go test
+// It can be used for Fuzz testing with go test -fuzz=FuzzQuery
+// These tests can be extended in 3 ways;
+// a. Add more sample data to the seed-data.test file
+// b. Add more sample queries to the seed-queries.test file
+// c. Add additional step sizes or time ranges to consider
+// When the Go Fuzzer is used, it will randomize the inputs of query string, start time, end time and step size.
+func FuzzQuery(f *testing.F) {
+
+	steps := []time.Duration{
+		time.Second,
+		time.Second * 2,
+		time.Second * 10,
+		time.Second * 30,
+		time.Second * 59,
+		time.Minute,
+		time.Second * 61,
+		time.Minute * 10,
+		time.Hour * 24 * 365 * 100,
+	}
+
+	epoch := time.Unix(0, 0)
+
+	ranges := [][]time.Time{
+		// invalid time range, as start > end
+		{epoch.Add(time.Second), epoch},
+
+		// start == end
+		{epoch, epoch},
+
+		// explore below our sample interval
+		{epoch, epoch.Add(time.Second * 1)},
+
+		// explore the bounds around our sample interval
+		{epoch, epoch.Add(time.Second * 59)},
+		{epoch, epoch.Add(time.Second * 60)},
+		{epoch, epoch.Add(time.Second * 61)},
+		{epoch.Add(time.Second * 4), epoch.Add(time.Second * 61)},
+
+		// explore the bounds after we have multiple samples
+		{epoch, epoch.Add(time.Minute*5 - time.Second)},
+		{epoch, epoch.Add(time.Minute * 5)},
+		{epoch, epoch.Add(time.Minute*5 + time.Second)},
+		{epoch.Add(time.Second * 4), epoch.Add(time.Second * 61)},
+		{epoch.Add(time.Minute * 2), epoch.Add(time.Second * 61)},
+	}
+
+	// For each query string we will test for each range & step combination
+	// The fuzzer will then be able then randomise on the query, time range and/or step size
```

Comment:
Maybe something like this?
```suggestion
	// Generate Fuzzing arguments with different ranges and steps combination for the passed query.
```